lib/passport-saml/saml.js

```diff
@@ -689,6 +680,150 @@ SAML.prototype.validatePostResponse = function (container, callback) {
   .done();
 };
 
+SAML.prototype.validateInResponseTo = function(inResponseTo) {
+  if(this.options.validateInResponseTo){
+    return Q.ninvoke(this.cacheProvider, 'get', inResponseTo)
+      .then(function(result) {
+        if (!result)
+          throw new Error('InResponseTo is not valid');
+        return Q();
+      });
+  } else {
+    return Q();
+  }
+}
+
+SAML.prototype.validateRedirect = function(container, callback) {
+  var self = this;
+  var type = container.SAMLRequest ? 'SAMLRequest' : 'SAMLResponse';
+
+  var data = new Buffer(container[type], "base64");
+  zlib.inflateRaw(data, function(err, inflated) {
+    if (err) {
+      return callback(err);
+    }
+
+    var dom = new xmldom.DOMParser().parseFromString(inflated.toString());
+    var parserConfig = {
+      explicitRoot: true,
+      tagNameProcessors: [xml2js.processors.stripPrefix]
+    };
+    var parser = new xml2js.Parser(parserConfig);
+    parser.parseString(inflated, function (err, doc) {
+      if (err) {
+        return callback(err);
+      }
+
+      Q.fcall(function () {
+        return type === 'SAMLResponse' ? self.verifyLogoutResponse(doc) : self.verifyLogoutRequest(doc);
+      })
+      .then(function() {
+        var signature = container.Signature;
+        if (signature) {
+          if (self.options.cert) {
+            return self.certsToCheck()
+              .then(function(certs) {
+                var urlString = type + '=' + encodeURIComponent(container[type]);
+
+                if (container.RelayState) {
+                  urlString += '&RelayState=' +
+                    encodeURIComponent(container.RelayState);
+                }
+
+                urlString += '&SigAlg=' + encodeURIComponent(container.SigAlg);
+
+                var hasValidQuerySignature = certs.some(function (cert) {
+                  return validateSignatureForRedirect(
+                    urlString, signature, container.SigAlg, cert
+                  );
+                });
+
+                if (!hasValidQuerySignature) {
+                  throw 'Invalid signature';
+                }
```

Comment:
Should we return a `Q()` here as well to have a common method return value?

Reply:
I threw an error so it will resolve with fail